### ISSUE TITLE
Allow to differentiate configuration for project from configuration for other code

### ISSUE THREAD

**Issue by gmazzap** [vimeo/psalm#3250]
This issue was created after #3226

-----

Long-story short, I need to integrate a 3rd party library that does very nasty things with doc blocks<sup>1</sup>.

So I decided best option would be to set `useDocblockTypes="false"`.

This worked quite well (my code uses type declarations close to everywhere so no big deal).

The problem started when I wanted to use Pslam templates. Expecially templates with constraints. As said in #3226 those really require doc blocks.

My idea to solve this issue is that maybe would be very useful, not only for just this issue, to be able to configure a different set of options for _project_ code and _included_ code.

My idea would be to have "scopes" for configuration.

Imagine, for example, a configuration like this:

```xml
<psalm totallyTyped="true" useDocblockTypes="true">

	<projectFiles>
		<directory name="src"/>
	</projectFiles>

	<scopes>

		<scope totallyTyped="false" useDocblockTypes="false">
			<scopeFiles>
				<directory name="vendor/foo/bar"/>
			</scopeFiles>
			<issueHandlers>
				<MixedAssignment errorLevel="suppress"/>
			</issueHandlers>
		</scope>

		<scope totallyTyped="false">
			<scopeFiles>
				<directory name="vendor/one"/>
				<directory name="vendor/two"/>
			</scopeFiles>
		</scope>

	</scopes>
</psalm>
```

So we have a `<scopes>` node that contains a collection of `<scope>` nodes.

Each of these  `<scope>` would define a target of files/directories in the same way root `<psalm>` does, but using `<scopeFiles>` instead of `<projectFiles>` and could contain _almost_ anything that the _root_ `psalm` node can contain.

(For example, would make no sense to allow `<projectFiles>` inside `<scope>`).

When analyzing a file, Psalm would check if it belongs to any of the defined scope and if so it would use a configuration obtained _merging_ the _overall_ configuration with the scope-specific configuration.

This approach would allow a lot of flexibility in the way code is analyzed when dealing with 3rd party code.

Note that this is just a raw, not-fully-thinked approach, created with very little knowledge on the internal of Psalm, so consider it more an idea in the regard of _"having different settings for 3rd party libraries"_ than a request including implementation details.

-----

<sup>1</sup> _A non exhaustive list: saying it returns array where it returns object, omit `|null` where it returns nullable, saying that a param must be of a kind, but actually requiring another type and so on...
All of this without using any type declaration of sort._

**Comment by muglug** [comment]
This seems far out of scope, sorry. I'd recommend exploring Psalter's ability to perform bulk docblock updates.
